pkg/apps/apis/apps/validation/validation.go

```diff
@@ -49,32 +52,68 @@ func ValidateDeploymentConfigSpec(spec appsapi.DeploymentConfigSpec) field.Error
 		allErrs = append(allErrs, kapivalidation.ValidateNonnegativeField(int64(*spec.RevisionHistoryLimit), specPath.Child("revisionHistoryLimit"))...)
 	}
 	allErrs = append(allErrs, kapivalidation.ValidateNonnegativeField(int64(spec.MinReadySeconds), specPath.Child("minReadySeconds"))...)
+
 	timeoutSeconds := appsapi.DefaultRollingTimeoutSeconds
 	if spec.Strategy.RollingParams != nil && spec.Strategy.RollingParams.TimeoutSeconds != nil {
 		timeoutSeconds = *(spec.Strategy.RollingParams.TimeoutSeconds)
 	} else if spec.Strategy.RecreateParams != nil && spec.Strategy.RecreateParams.TimeoutSeconds != nil {
 		timeoutSeconds = *(spec.Strategy.RecreateParams.TimeoutSeconds)
 	}
+
 	if timeoutSeconds > 0 && int64(spec.MinReadySeconds) >= timeoutSeconds {
 		allErrs = append(allErrs, field.Invalid(specPath.Child("minReadySeconds"), spec.MinReadySeconds, fmt.Sprintf("must be less than the deployment timeout (%ds)", timeoutSeconds)))
 	}
+
 	if spec.Strategy.ActiveDeadlineSeconds != nil && int64(spec.MinReadySeconds) >= *(spec.Strategy.ActiveDeadlineSeconds) {
 		allErrs = append(allErrs, field.Invalid(specPath.Child("minReadySeconds"), spec.MinReadySeconds, fmt.Sprintf("must be less than activeDeadlineSeconds (%ds - used by the deployer pod)", *(spec.Strategy.ActiveDeadlineSeconds))))
 	}
+
 	if spec.Template == nil {
 		allErrs = append(allErrs, field.Required(specPath.Child("template"), ""))
 	} else {
 		originalContainerImageNames := getContainerImageNames(spec.Template)
 		defer setContainerImageNames(spec.Template, originalContainerImageNames)
 		handleEmptyImageReferences(spec.Template, spec.Triggers)
 		allErrs = append(allErrs, validation.ValidatePodTemplateSpecForRC(spec.Template, spec.Selector, spec.Replicas, specPath.Child("template"))...)
+
 	}
+
 	if spec.Replicas < 0 {
 		allErrs = append(allErrs, field.Invalid(specPath.Child("replicas"), spec.Replicas, "replicas cannot be negative"))
 	}
+
 	if len(spec.Selector) == 0 {
 		allErrs = append(allErrs, field.Invalid(specPath.Child("selector"), spec.Selector, "selector cannot be empty"))
+	} else {
+		// Validate reserved selectors
+
+		// DC.Spec.Selector[appsapi.DeploymentConfigLabel] shall always equal to the DC.Name
+		// If left empty Canonicalization will set it to DC.Name
```

Comment:
if we're going to force this, would it be better to set it in PrepareForCreate, so validation can observe the set value?

Reply:
(and maybe PrepareForUpdate?)

Reply:
I think [Canonicalize](https://github.com/openshift/origin/blob/c07d935e1a000e89ec84dbc70683c797c1c421a8/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L61-L67) is the right place for it. It translates DC into canonical form where the implied selector is expressed explicitly but those objects are still equivalent. Runs before validation.

Validation should validate user input, not a canonicalized object - the result of validation might be confusing in such case.

Reply:
> It translates DC into canonical form where the implied selector is expressed explicitly but those objects are still equivalent.

The implicit selector was in the behavior of another component, not in the API object itself. As long as it is based on an immutable attribute of the object (the name), I guess it is ok to do this way.

> Validation should validate user input, not a canonicalized object - the result of validation might be confusing in such case.

Alternately, putting it here means we have no check on the computed value. It is unnerving to me to have changes made to an object after validation.

Reply:
@liggitt are you saying that what I do is not canonicalization or that canonicalization is implemented wrongly in apiserver?

Reply:
my main two questions/objections are:
* if you have an existing object that fails this new validation (has a conflicting name label), will that prevent the no-op updates we do on upgrade? if so, we have to adjust the validation to not fail an update of an already-invalid deployment config.
* why is it better to do the defaulting of the name label after validation instead of in PrepareForCreate/PrepareForUpdate, where we have a validation pass to catch any errors?

> Validation should validate user input, not a canonicalized object - the result of validation might be confusing in such case.

Validation should prevent invalid data from being persisted to etcd, no matter the source. If we have a bug in our server logic, I'd rather it prevent persistence than get stored and need to be migrated later.

Reply:
> will that prevent the no-op updates we do on upgrade

Not sure what you mean by that.

> why is it better to do the defaulting of the name label after validation instead of in PrepareForCreate/PrepareForUpdate, where we have a validation pass to catch any errors?

Because that's the way how it's implemented in apiserver - [Canonicalize](https://github.com/openshift/origin/blob/c07d935e1a000e89ec84dbc70683c797c1c421a8/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L61-L67) is run after validation.

I think that either I am not doing canonicalization here or you want canonicalization to be run at different place. Helping to identify which one it is will target as in the right direction.

Judging by your concerns about changes after validation you likely want [Canonicalize](https://github.com/openshift/origin/blob/c07d935e1a000e89ec84dbc70683c797c1c421a8/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L61-L67) to be run in different place, before validation.

Reply:
>> will that prevent the no-op updates we do on upgrade
>
> Not sure what you mean by that.

If an existing invalid object is persisted, and this tightened validation is added (that disallows a non-standard name label), what will happen when we upgrade and try to get/put the newly-invalid object as part of the upgrade (before your controller fixup migration that stomps the labels and selector has run)?

> I think that either I am not doing canonicalization here or you want canonicalization to be run at different place. Helping to identify which one it is will target as in the right direction.

I think it is semantics whether moving implicit behavior from the controller into the API object is canonicalization or not. I just want the data going into etcd to be validated.

Reply:
> If an existing invalid object is persisted, and this tightened validation is added (that disallows a non-standard name label), what will happen when we upgrade and try to get/put the newly-invalid object as part of the upgrade (before your controller fixup migration that stomps the labels and selector has run)?

In that brief period after the validation is deployed and before the new version of the controller fixes all the DCs with invalid labels, you can update the object only by making it valid manually, validation will tell you what field is causing it.
If there is an action to be taken on a DC by ansible as part of the ugrade, it should be fine as we shouldn't have our own invalid DCs - which also will be tested by upgrade on the testing clusters.

> I think it is semantics whether moving implicit behavior from the controller into the API object is canonicalization or not. I just want the data going into etcd to be validated.

So take this in general. Every canonicalization is a subtle change of an object leaving it still "equivalent" to what it was before but binary different. Say you canonicalize a value from 0.0 to 0. The way it is implemented now in apiserver, [Canonicalize](https://github.com/openshift/origin/blob/c07d935e1a000e89ec84dbc70683c797c1c421a8/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L113) runs after validation. So by definition here you are writing to etcd something that wasn't validated in it's final form. Either that is intentional or a bug.

I don't really feel comfortable creating a new method CanonicalizeBeforeValidation if there is already Canonicalize interface method dealing with it. If that method is called in the wrong place and it's result should go though validation we should probably fix it, if not we should likely use it.

I think the reason to do this after validation is the fact that otherwise you would act on unvalidated data so you would duplicate all the branching logic (checks) already in validation. So the only way would be to validate it once again. I think that we can make sure with tests that canonicalization won't result in invalid object. (Here is one instance of it https://github.com/tnozicka/origin/blob/3fd223af21a17d6e322c9120e6a89eb14ed4aa27/pkg/apps/registry/deployconfig/strategy_test.go#L48-L50)

Maybe the answer is to run validation twice to make really sure it's valid before canonicalize and after. Either way I tend to feel that it should be solved in apiserver, I am just implementing an interface (method) here.

Reply:
> In that brief period after the validation is deployed and before the new version of the controller fixes all the DCs with invalid labels, you can update the object only by making it valid manually, validation will tell you what field is causing it.

We have automatic no-op updates that run during the install process. This validation change on update will break that process, and break our upgrade.

Reply:
> So by definition here you are writing to etcd something that wasn't validated in it's final form. Either that is intentional or a bug.

We have exactly one use of Canonicalize today. I don't think the current placement was well considered. I'm really hesitant to expand it using it on a field validated in such a complex way.

What's the downside of this approach:
* set the name in the selector in PrepareForCreate and and PrepareForUpdate
* validate the selector on create
* validate the selector on update if the old object passed validation
* do the controller migration to set or override the name in the selector

Reply:
> We have exactly one use of Canonicalize today.

and 1 [upstream](https://github.com/tnozicka/origin/blob/710998e7358752a4c79518804522f1a6cf239838/vendor/k8s.io/kubernetes/pkg/registry/core/endpoint/strategy.go#L58-L62)

Reply:
> Ok, that seems serious. How is that handled w.r.t. to need for manual migrate command today? Would you mind sharing a link to where we do that?

Manual migration is typically for data fixup, not to pass stricter validation. Storage migration (get/put of all API objects) is done before and after upgrade with the `oc adm migrate storage` command:
https://github.com/openshift/openshift-ansible/blob/master/playbooks/openshift-master/private/upgrade.yml

> I guess every backwards incompatible change in validation must have potentially broken this process.

Yes.

> How were we solving it in the past?

By preventing backwards incompatible validation changes with constant vigilance upstream and reverting them when we find them (which we inevitably do on upgrade).